Track Clang function types in the AST

varungandhi-apple opened the pull request:
The 3rd commit introduces a bunch of TODOs, most of which are resolved in the 4th commit.

Things that are implemented:
1. ClangTypeConverter is available at the AST level. It is used both when resolving types (TypeResolution) and when building types (ASTBuilder).
2. Including the clang types in the type equality tests.

Questions:
1. ~It is not clear who/what should be maintaining ownership of the ClangTypeConverter in the ASTBuilder. There are a couple of TODOs related to this. ("Is this too expensive?").~
2. ~How should I further be testing this change?~

Things that are not implemented yet but I plan to do as part of this PR (unless someone says I should make a follow-up PR about it):
- [x] Test cases -- ~right now, the main "test" is the whole test suite passes, i.e. we don't mess up type equality anywhere.~ added some trickier test cases
- [x] ~Setting the clang types in SILFunctionType.~ (punted to future PR)
- [x] ~Reporting differences in the clang type in diagnostics.~ (punted to future PR)

Since it has taken me much more time than anticipated to get to this point, I figured I'd try to get feedback sooner rather than later (even though a bunch of things are not implemented), especially given that this is a relatively large PR 😐.

varungandhi-apple committed:
[NFC] Clean up GenClangType a little bit.

varungandhi-apple committed:
[NFC] Check the cache first when computing Clang types.

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
Ah, there is a compilation failure to the removed `Builtin.UnknownObject`.

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
Ah we are crashing when compiling SwiftPM :from

varungandhi-apple commented:
We're compiling SwiftPM properly on Linux but not on OS X? 😦

rjmccall commented on include/swift/AST/Types.h:
```diff
@@ -39,6 +39,25 @@
+#ifndef NDEBUG
+#include "clang/AST/Type.h"
+namespace {
+void checkIsFunctionType(const clang::Type *type) {
```
We usually just declare helper functions like this `static` (`static inline` since this is a header) instead of putting them in an anonymous namespace.  But actually I don't think it's a good idea to define this in the header, *especially* since it conditionally includes a major header, which exposes us to all sorts of risk where code will compile in NDEBUG builds but not otherwise; just define it out-of-line.  Also consider making it a static function on ExtInfo, which is the only place that uses it.

rjmccall commented on include/swift/AST/Types.h:
```diff
@@ -2844,39 +2863,62 @@ class AnyFunctionType : public TypeBase {
+    struct Uncommon {
+      // We preserve a full clang::Type *, not a clang::FunctionType *, so
+      // we can keep sugar in case we need to present an error to the user.
+      const clang::Type *ClangFunctionType;
```
This is what we want for the formal type `AnyFunctionType`, but it probably shouldn't be true of `SILFunctionType`, which is always canonical and therefore should probably be storing a canonical Clang type as well.  That should at least be explained here, since you're reusing the type in both places.

varungandhi-apple replied:
> That should at least be explained here, since you're reusing the type in both places.

Ah, the name is the same in both places but the structure is different. Right now, I'm storing a `clang::FunctionType *` in `SILFunctionType`, but yeah, I should store a canonical type instead.

varungandhi-apple replied:
We decided to use `clang::FunctionType *` to avoid a dependency on the clang's Type.h in Types.h. We will maintain the invariant that it is canonical by hand.

rjmccall commented on include/swift/AST/Types.h:
```diff
@@ -2844,39 +2863,62 @@ class AnyFunctionType : public TypeBase {
+      // we can keep sugar in case we need to present an error to the user.
+      const clang::Type *ClangFunctionType;
+
+      bool isNull() const { return !ClangFunctionType; }
```
`empty()`, maybe?

rjmccall commented on lib/IRGen/GenClangType.cpp:
```diff
@@ -475,8 +475,8 @@ GenClangType::visitBoundGenericType(CanBoundGenericType type) {
-  auto loweredArgTy = IGM.getLoweredType(args[0]).getASTType();
-
+  auto argSILType = IGM.getLoweredType(args[0]);
+  auto argLoweredTy = argSILType.getASTType();
```
I don't understand why you've changed this variable name; it seems spurious.

varungandhi-apple replied:
This change was made before the update to the getASTType documentation. Forgot to remove it.

rjmccall commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -2709,7 +2709,8 @@ static SILFunction *getOrCreateKeyPathGetter(SILGenModule &SGM,
                              /*pseudogeneric*/ false,
-                             /*noescape*/ false),
+                             /*noescape*/ false,
+                             /*clangFunctionType*/ nullptr),
```
This specific pattern (thin with no unusual properties) seems to be common enough in SILGen that there probably ought to be a `getThin()` convenience factory for it on `ExtInfo`.  Whether that makes sense e.g. for the uses in ConstraintSystem is a separate question.

varungandhi-apple replied:
Oops, forgot to address this. 🤦‍♂  On it.

rjmccall commented on include/swift/AST/ClangTypeConverter.h:
```diff
@@ -0,0 +1,112 @@
+      std::tuple<ASTContext &, clang::ASTContext &, llvm::Triple>;
+
+  /// Static helper to avoid duplication at the constructor's call-site.
+  static ConstructorArgs getConstructorArgs(ASTContext &ctx);
```
Why this abstraction instead of just taking an `ASTContext &` and then pulling the components out in the constructor?

varungandhi-apple replied:
IIUC, reference data members (such as `clang::ASTContext &ClangASTContext`) need to be set in the initializer list in the constructor. This means that I'd have to write a lambda which returns the `clang::ASTContext &` and duplicate most of that logic for initializing the `llvm::Triple` field (either in the initializer list or in the body of the constructor).

rjmccall replied:
You can use previously-initialized fields in the initializers for later fields.

varungandhi-apple commented on include/swift/AST/ASTDemangler.h:
```diff
@@ -24,6 +24,7 @@
 
 #include "llvm/ADT/ArrayRef.h"
 #include "llvm/ADT/StringRef.h"
+#include "swift/AST/ClangTypeConverter.h"
```
> Why does ClangTypeConverter need to exist as public API, anyway? I would expect there to just be some sort of query, maybe even a request, either on ASTContext or Type.

It probably does not. I just went for the simplest thing first 😅. Will look at how we can requestify it.

varungandhi-apple replied:
I've made it a simple method for now, I'll make it a request in a later PR if we think that'd be desirable.

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
The crash is because of a type mismatch issue, we have a Clang type in one case and none in the other. Think I should put a more descriptive crash message there instead of just an `llvm_unreachable`.

varungandhi-apple commented:
Minimal test case. Only `SIG_DFL` is problematic, other signals seem to work ok.

```
// RUN: %target-swift-frontend -typecheck %s
// REQUIRES: OS=macosx

import Darwin

func f() {
  let s: sig_t? = SIG_DFL
}
```

varungandhi-apple commented:
Think I understand what the issue is. It has nothing to do with SIG_DFL and probably has everything to do with (de)serialization. We're not (de)serializing Clang types yet, so decls loaded from a different module don't have it. 🤦‍♂

The only way I see is disabling the clang function pointer check when comparing types.

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
I still need to add more test cases utilizing the new flag, but the core logic should be good now.

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
The LLDB failure was fixed here: https://github.com/apple/llvm-project/pull/24

@swift-ci please test

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
@swift-ci please test source compatibility

varungandhi-apple commented:
@swift-ci Please Test Source Compatibility

rjmccall commented on include/swift/AST/ASTContext.h:
```diff
@@ -561,6 +562,15 @@ class ASTContext final {
+  /// \param repr The Swift calling convention, which must be C-compatible.
+  const clang::Type *
+  getClangFunctionType(Type resultTy, ArrayRef<Type> paramsTy,
+                       FunctionTypeRepresentation repr);
```
Can you make these arguments exactly match the normal inputs to `getFunctionType`, so that it takes an array of `AnyFunctionType::Param` and an `ExtInfo` (albeit one with either the wrong representation or (obviously) a missing Clang type)?

varungandhi-apple replied:
> Can you make these arguments exactly match the normal inputs to getFunctionType, so that it takes an array of AnyFunctionType::Param and an ExtInfo

I'm guessing you meant `swift::FunctionType::get` here? Could you point out why this is beneficial?

The order of the arguments is written so as to match the Clang convention (result type first) instead of the Swift convention (params first). I suppose that is a bit questionable, since we are passing Swift types in...

> albeit one with either the wrong representation or (obviously) a missing Clang type

Both of these options are unappealing IMO because (a) the first one is just wrong and (b) the second one means we won't be able to add the invariant in the future that we always have a clang type. I want to ensure correctness-by-construction so that we have less confusing moments when debugging like "oh, this is in an incorrect/invalid state here but there's a good reason for that".

rjmccall replied:
I don't care about the exact order of arguments.

You should take an array of `Param`s instead of `Type`s because we very well might want to generate a different Clang type based on the parameter attributes.  It also looks like it'll generally be easier for the call sites.

The argument for ExtInfo is similar, in that we might well want to change the Clang function type based on Swift function-level attributes.  For example, we currently don't allow `@convention(c)` function types to be throwing, but if we did, it would certainly affect the exported Clang function type.  Experience suggests that the set of function attributes tends to grow over time.  I agree with your argument against allowing the construction of an ExtInfo with C representation and no type, but I think you can reasonably pass down an ExtInfo with a comment that the representation is meaningless and then pass down the real representation separately.

varungandhi-apple replied:
Got it, thanks for explaining!

rjmccall commented on include/swift/AST/Types.h:
```diff
@@ -2872,27 +2878,59 @@ class AnyFunctionType : public TypeBase {
+    }
+
+    ExtInfo(unsigned Bits, const clang::Type *type)
+      : ExtInfo(Bits, Uncommon(type)) {}
```
Probably not worth having a private convenience constructor.

rjmccall commented on include/swift/AST/Types.h:
```diff
@@ -2944,25 +2987,28 @@ class AnyFunctionType : public TypeBase {
+    std::pair<unsigned, const void *> getFuncAttrKey() const {
+      const void *ptr = nullptr;
+      if (Other.ClangFunctionType)
+        ptr = Other.ClangFunctionType;
```
`const void *ptr = Other.ClangFunctionType;`?  Or if you're trying to let `Other` abstract itself, just give it a method.

varungandhi-apple replied:
Ah, I had some printing logic here for debugging, forgot to clean up the code later.

rjmccall commented on include/swift/AST/Types.h:
```diff
@@ -2993,11 +3039,13 @@ class AnyFunctionType : public TypeBase {
   : TypeBase(Kind, CanTypeContext, properties), Output(Output) {
-    Bits.AnyFunctionType.ExtInfo = Info.Bits;
+    Bits.AnyFunctionType.ExtInfoBits = Info.Bits;
+    Bits.AnyFunctionType.HasUncommonInfo = Info.getUncommonInfo().hasValue();
```
Please leave a comment saying that subclasses are responsible for initializing their uncommon-info storage.

rjmccall commented on include/swift/AST/Types.h:
```diff
@@ -3035,8 +3083,21 @@ class AnyFunctionType : public TypeBase {
+    return ExtInfo(Bits.AnyFunctionType.ExtInfoBits, getClangFunctionType());
+  }
+
+  ExtInfo getCanonicalExtInfo(bool useClangFunctionType) const {
```
Please include a comment explaining that this parameter is only for staging purposes while we teach Swift to provide function types reliably.

rjmccall commented on include/swift/AST/Types.h:
```diff
@@ -3117,6 +3187,21 @@ class FunctionType final : public AnyFunctionType,
   }
 
+  const clang::Type *getClangFunctionType() const {
+    switch (numTrailingObjects(OverloadToken<ExtInfo::Uncommon>())) {
```
?

varungandhi-apple replied:
I'm guessing the ? was for "why is this implementation unnecessarily complicated"; fixed that.

rjmccall replied:
Thanks!  Looks a lot better.

rjmccall commented on include/swift/Basic/LangOptions.h:
```diff
@@ -257,6 +257,11 @@ namespace swift {
 
+    /// Use Clang function types for computing canonical types.
+    /// If this option is false, the clang function types
+    /// This option should be turned on once we have serialization and
```
Please finish this comment.

rjmccall commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -2498,19 +2496,35 @@ Type TypeResolver::resolveASTFunctionType(FunctionTypeRepr *repr,
   if (auto genericEnv = repr->getGenericEnvironment()) {
     outputTy = outputTy->mapTypeOutOfContext();
+    FunctionType::ExtInfo extInfo(
+      representation, noescape, repr->throws(), /*clangFunctionType*/ nullptr);
```
Please continue computing the `extInfo` in a common location in this function.

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
@swift-ci please test

rjmccall commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -0,0 +1,684 @@
+    return clangCtx.getObjCObjectPointerType(clangType);
+  }
+
+  return getClangIdType(ClangASTContext);
```
This doesn't actually work for a non-ObjC existential, and I don't think they're generally bridgeable, either.

varungandhi-apple replied:
Does this mean that the original code had a bug, which just didn’t get triggered?

rjmccall replied:
Probably just assuming that this is only ever exercised with specific types.

varungandhi-apple replied:
Removing this (and crashing instead) causes a couple of tests to fail:

```
Interpreter/SDK/objc_mangling.swift
sodlib/SwiftObjectNSObject.swift
```

Both of those failures are for %target-run lines. I don't want to spend time into digging deeper here right now. Will leave a TODO.

rjmccall commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -0,0 +1,684 @@
+ClangTypeConverter::visitBoundGenericClassType(BoundGenericClassType *type) {
+  // Any @objc class type in Swift that shows up in an @objc method maps 1-1 to
+  // "id <SomeProto>"; with clang's encoding ignoring the protocol list.
+  return getClangIdType(ClangASTContext);
```
We should probably at least try to preserve type arguments on imported ObjC generic classes.

varungandhi-apple replied:
I tried implementing this based on the other things we already have but it didn't work correctly (convert generic args, cast result, use quals() to get protocol decls, use that to create another ObjCObjectType with id as the base).

This doesn't seem super important right now, leaving a TODO instead.

rjmccall commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -0,0 +1,684 @@
+clang::QualType
+ClangTypeConverter::visitBoundGenericType(BoundGenericType *type) {
+  // The only possibilities are *Pointer<T>, SIMD*<T>,
+  // ImplicitlyUnwrappedOptional<T> and Optional<T>.
```
This is another point where it's interesting to ask whether this function should be allowed to assume that the type is exportable to C as a precondition.  I'm almost positive that there are APIs for asking whether a type is any of these standard library types.

varungandhi-apple replied:
> This is another point where it's interesting to ask whether this function should be allowed to assume that the type is exportable to C as a precondition.

Strictly speaking, it doesn't assume that (unless I made a mistake). If you look at the first item `switch (kind) { case StructKind::Invalid: `, in `GenClangType` we have a `llvm_unreachable` whereas in `ClangTypeConverter` we silently fail with `return clang::QualType()`.

This is important because someone might make a mistake in typing something and we should be issuing a diagnostic instead of crashing. I've replaced other similar crash points in `GenClangType` with `return clang::QualType()` as well.

rjmccall replied:
Hmm, okay.

I do think you probably at least need a bridged-to-ObjC check here.

rjmccall commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -0,0 +1,684 @@
+ClangTypeConverter::visitProtocolCompositionType(ProtocolCompositionType *type) {
+  // Any will be lowered to AnyObject, so we return the same result.
+  if (type->isAny())
+    return getClangIdType(ClangASTContext);
```
I'd forgotten that we're doing both bridging and translation in one step here.  Okay.

varungandhi-apple replied:
I don't understand... aren't bridging and translation synonymous? What's the difference?

rjmccall replied:
The Swift type `Swift.Any` is bridged to the Swift type `Swift.AnyObject`, which can be translated to the Clang type `id`, which it is assumed to have identical ABI treatment to.

varungandhi-apple commented on include/swift/AST/Types.h:
```diff
@@ -2887,27 +2893,57 @@ class AnyFunctionType : public TypeBase {
+
+    static const clang::Type *getClangFunctionType(const Uncommon *uncommon) {
+      return uncommon->ClangFunctionType;
+    }
```
It was being used earlier, because `AnyFunctionType` was a friend of `ExtInfo::Uncommon`, but not `FunctionType` was not a friend. After I made `FunctionType` a friend, I forgot to delete this method.

I wasn't really sure how "private" the `Uncommon` field oughta' be 😐.

varungandhi-apple commented on lib/AST/ASTContext.cpp:
```diff
@@ -4340,6 +4360,25 @@ Type ASTContext::getBridgedToObjC(const DeclContext *dc, Type type,
+    impl.Converter.emplace(
+      *this,
+      cml->getClangASTContext(),
+      llvm::Triple(cml->getTargetInfo().getTargetOpts().Triple));
```
I don't think that's quite right, but I'm replacing it with `LangOpts.Target`.

varungandhi-apple commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -0,0 +1,684 @@
+static clang::QualType getClangVectorType(const clang::ASTContext &ctx,
+                                          clang::BuiltinType::Kind eltKind,
+                                          clang::VectorType::VectorKind vecKind,
+                                          StringRef numEltsString) {
```
This whole function looks a bit fishy. I don't get why there is a `(void) failedParse` or that seemingly unnecessary template parameter (maybe a compiler inference bug earlier?).

> Can we just look through these types down to the underlying BuiltinVectorType? I guess this must've been existing code, but it seems really unfortunate.

I don't get why "it seems really unfortunate" but yeah, this is basically the same as the existing code in `GenClangType.cpp`, warts included. I've tried to keep non-functional divergences between the two files to a minimum.

rjmccall replied:
Well, it's unfortunate that we're parsing the name of a public library type in order to break it down into a vector component and length that in theory we could recover in some other way.

varungandhi-apple replied:
Another case where if C++ had enums cases with associated values, this would've been a non-issue. Leaving a TODO for now.

varungandhi-apple commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -0,0 +1,684 @@
+clang::QualType
+ClangTypeConverter::visitExistentialMetatypeType(ExistentialMetatypeType *type) {
+  return getClangMetatypeType(ClangASTContext);
+}
```
> How important do we think reliable rejection is for this API? This implementation is definitely converting some types that don't actually have ObjC equivalents.

I don't know 😐. I realized there is an undocumented assumption here which you're kinda' calling out. The contract is:

> We will compute the right Clang type if the Swift type has a corresponding Clang type

not

> We will compute the right Clang type if and only if the Swift type has a corresponding Clang type

We are relying on type-checker and its friends to reject ill-formed`@convention(c)` stuff.

Another somewhat related thing that worries me a bit is that because of the code split (e.g. `ClangTypeConverter` doesn't handle `SILFunctionType` and the like whereas `GenClangType` does), we won't be able to eventually combine the two converters together. Do these two need to have different contracts?

rjmccall replied:
Remember that the `GenClangType` converter literally won't exist in a month when we can rely on the Clang function type being present.

rjmccall replied:
As for the general question, I think it would be nice to make as much of this code as possible try to do the right thing (i.e. return null) for types without a corresponding Clang type, and then we can call out the places where that's difficult or expensive.

varungandhi-apple replied:
Leaving TODOs here instead of adding stronger checking right now.

varungandhi-apple commented:
(Testing partial work, haven't finished addressing all the review comments.)

@swift-ci please test

varungandhi-apple commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -0,0 +1,684 @@
+    auto *PDecl = clang::ObjCProtocolDecl::Create(
+                    const_cast<clang::ASTContext &>(clangCtx),
+                    clangCtx.getTranslationUnitDecl(), name,
+                    clang::SourceLocation(), clang::SourceLocation(), nullptr);
```
1. How would I check the import?
2. “Re-use the fake protocol” — won’t that happen automatically because ClangTypeConverter caches things anyways? Or is this about the fact that we don’t cache sub-computations?

rjmccall replied:
We discussed this privately, but just for a public answer: for the first, you can check whether the Swift Decl has a clang node, which should be a `clang::ObjCProtocolDecl*`; for the second, you should endeavor to make sure that different Swift types use the same Clang declaration, both to keep down the memory use and to make this more useful as we find more clients.

varungandhi-apple commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -0,0 +1,688 @@
+    if (args.size() != 1)
+      // We have a weird Optional with multiple generic arguments, eh?
+      return clang::QualType();
+    return convert(args[0]);
```
@rjmccall what should we be doing here? This seems very wrong, but I wasn't sure what to do. We cannot simply reuse the code from `GenClangType`.

```
  if (auto underlyingTy =
          SILType::getPrimitiveObjectType(type).getOptionalObjectType()) {
    // The underlying type could be a bridged type, which makes any
    // sort of casual assertion here difficult.
    return Converter.convert(IGM, underlyingTy.getASTType());
  }
```

rjmccall replied:
The only Optional types that have Clang equivalents are those with single-class-reference representations; there's a method around somewhere for that.

varungandhi-apple replied:
> The only Optional types that have Clang equivalents are those with single-class-reference representations; there's a method around somewhere for that.

I found `ImportName.cpp/canImportAsOptional` which is probably the one 🙂. There's also `DeclAndTypePrinter::isClangPointerType` which has a similar check.

varungandhi-apple commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -0,0 +1,688 @@
+    auto clangType  = clangCtx.getObjCInterfaceType(CDecl);
+    return clangCtx.getObjCObjectPointerType(clangType);
+  }
+  return getClangIdType(clangCtx);
```
@rjmccall I'm guessing this has the same issues as with the `visitProtocolType` method? (Not caching decls and returning a bogus `id` for non-ObjC stuff)

rjmccall replied:
Github lost track of where this is after the rebase.  If this is the protocol-composition method, yeah, ideally they would re-use a common method for translating a `ProtocolDecl`.  If this is the class method, then yeah, same principle: we should cache the "exported" `ObjCInterfaceDecl`.

varungandhi-apple replied:
Yes, this is in the `visitClassType` method. If I error out in `GenClangType` (which has similar code) if `!swiftDecl->isObjC()`, then I get a crash with

```
class SwiftClass {
      @objc func objcmethod() -> () {}
      func swiftmethod() -> () {}
      // Block attribute
      func f(_ someBlock: @convention(block) (Int64) -> Int64) {
      }
}
```

Not entirely sure why we're trying to convert that.

varungandhi-apple commented:
Still not done addressing all the review comments, but finished fixing conflicts when rebasing on top of the recent differentiable functions work.

@swift-ci please test

varungandhi-apple commented:
Ugh, rebasing suddenly created a bunch of whitespace changes 😭.

varungandhi-apple committed:
[AST] Setup AST and SIL to accomodate Clang function types.

We still don't store compute or store the type anywhere; we will do so in
later commits.

varungandhi-apple committed:
[AST] Add ClangTypeConverter, computing C types for FunctionTypes.

Note: The change in ASTBuilder::createFunctionType is functionally minor,
but we need the FunctionType::Params computed _before_ the ExtInfo, so we
need to shuffle a bunch of code around.

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
(Ugh, GitHub is showing the commits in chronological order instead of commit order. 🤦‍♂)

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple committed:
[NFC] Don't compute sizes manually in GenericFunctionType::get.

varungandhi-apple committed:
[NFC] Add some useful crash information in CSApply.

varungandhi-apple commented:
Touched the commits so that topological order matches chronological order and the UI is less confusing. _*sigh*_

@swift-ci please test

varungandhi-apple commented:
@swift-ci please test source compatibility

varungandhi-apple commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -0,0 +1,691 @@
+    // On 64-bit Windows, no C type is imported as an Int or UInt; CLong is
+    // imported as an Int32 and CLongLong as an Int64. Therefore, manually
+    // add mappings to C for Int and UInt.
+    if (Triple.isOSWindows() && Triple.isArch64Bit()) {
```
This needs to be changed to match the behavior in this recent PR. https://github.com/apple/swift/pull/20474

varungandhi-apple committed:
[NFC] Lift out common "canImportAsOptional" to a new header.

varungandhi-apple committed:
[NFC] Crash if we try to convert non-ObjC protocols.

This should be a non-functional change because we shouldn't be converting
protocols which aren't ObjC.

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
@swift-ci please test source compatibility

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
@swift-ci please test source compatibility

varungandhi-apple commented:
There is a merge conflict again, nooooooooo 😭

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
@swift-ci please test Linux platform

varungandhi-apple commented:
@swift-ci please smoke test Linux platform

varungandhi-apple commented:
@swift-ci please test Linux platform

varungandhi-apple commented:
The same Linux build failure seems to be happening with #28371 (build 17092) and #28374 (build 17094) as for us (builds 17093 and 17099).

varungandhi-apple commented:
@swift-ci please clean test Linux platform

varungandhi-apple commented:
@swift-ci please test Linux platform

varungandhi-apple commented:
Yasssss! Tests are passing. @rjmccall time for you to give this PR your blessings.

rjmccall commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -0,0 +1,728 @@
+  // ImplicitlyUnwrappedOptional<T> and Optional<T>.
+
+  if (!type->getDecl()->isObjC())
+    return clang::QualType();
```
Literally none of the above types are ObjC, so I'm curious how any of this works.

varungandhi-apple replied:
Modified the wrong function. I was working on `visitBoundGenericClassType`, which is right above this one.

varungandhi-apple replied:
As you point out, it almost certainly was not working, but we are not testing that yet. 😞 When I land the attribute, we can test all these out.

rjmccall commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -0,0 +1,728 @@
+clang::QualType
+ClangTypeConverter::visitBoundGenericType(BoundGenericType *type) {
+  // The only possibilities are *Pointer<T>, SIMD*<T>,
+  // ImplicitlyUnwrappedOptional<T> and Optional<T>.
```
Please remove `IUO` from this comment, as it doesn't exist anymore.

rjmccall commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -0,0 +1,728 @@
+
+    clang::QualType functionTy;
+    if (isa<SILFunctionType>(argCanonicalTy->getCanonicalType())) {
+      functionTy = convert(argCanonicalTy);
```
You actually assert on a `SILFunctionType` below.

varungandhi-apple replied:
Yes, that is intentional -- if something fails to convert, its visit method should have an error, instead of substituting the convert+visit method at all potential call-sites which could trigger the assertion.

rjmccall commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -0,0 +1,728 @@
+    return it->second;
+
+  // Try to do this without making cache entries for obvious cases.
+  if (auto nominal = type->getAs<NominalType>()) {
```
Oh, this check might obviate my comments above about handling the obvious cases.  If so, we should just assert that we aren't in those cases in those functions.

varungandhi-apple replied:
Added assertions.

varungandhi-apple commented on include/swift/AST/ClangSwiftTypeCorrespondence.h:
```diff
@@ -0,0 +1,36 @@
+//- ClangSwiftTypeCorrespondence.cpp - Relations between Clang & Swift types -//
```
> there's a standard form of these header comments

I ran out of characters to add `==` signs around.

varungandhi-apple replied:
If you have any suggestions for rephrasing, I can do that. This is the best I could come up with to fit in 80 chars.

rjmccall replied:
I think you're overthinking the filename and it could just be something like `ClangTypeInfo.h - Swift-related properties of Clang types`.

varungandhi-apple replied:
I think `Info` is very generic and doesn't hint at anything in particular.

varungandhi-apple commented on lib/AST/ASTContext.cpp:
```diff
@@ -41,6 +42,8 @@
 #include "swift/AST/SubstitutionMap.h"
 #include "swift/AST/SILLayout.h"
+#include "swift/AST/Type.h"
+#include "swift/AST/Types.h"
```
Yes it does, but it does so via other headers instead of explicitly. I'll delete it.

varungandhi-apple commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -0,0 +1,728 @@
+    return clang::QualType();
+
+  // TODO: As an optimization, we could check if proto has a clang decl
+  // already attached to it and if so, reuse it instead of making a new one.
```
The problem is the const-ness. You mentioned earlier that other things would need to be refactored to avoid a `const_cast` here. I didn't want to do that refactoring and I didn't want to do that `const_cast` either.

varungandhi-apple replied:
Ah never mind, we can replace this TODO with an assert because of the optimization.

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
@swift-ci please test

varungandhi-apple commented:
@rjmccall is this good to merge?

rjmccall commented:
See my comment in the other PR about `owned`, but we can do that in a follow-up; I think this is fine to merge now.
